### ISSUE TITLE
Type inference for dictionary Optional type value

### ISSUE THREAD

**Issue by falan** [microsoft/pyright#703]
**Describe the bug**
The behavior is the same with mypy`( error: Item "None" of "Optional[str]" has no attribute "strip")`, pyright`("strip" is not a known member of "None"Pyright (reportOptionalMemberAccess)`. But I am wondering it might be a bug because pyright already have the correct return type. 

**To Reproduce**
```python
i = {"a": "", "b": None}
dict_comp = {key: i[key].strip() if i[key] is not None else "" for key in i}
```
We have dict_comp with Dict[str, str] correctly. And there is no error for List  comprehension
```python

j = ["", None]
list_comp = [v.strip() if v is not None else "" for v in j]

```

**VS Code extension or command-line**
VS Code, pyright: 1.140

**Comment by erictraut** [comment]
Pyright (like mypy) assumes that subscript expressions (`[x]`) can return different type each time they are evaluated. Keep in mind that type checkers like Pyright and mypy don't track _values_ of expressions, they track _types_ of expressions. The type of the expression `i[x]` can depend on the _value_ of x. Consider the general case where the expression between the two brackets is a more complex expression like `i[random_number(x)]`.

One could argue that the expression `i[key]` is likely to return the same type if it is evaluated twice in succession with no reassignment of `key` in between, but that requires tracking _values_. I'm not aware of any type checkers in any language that handle this case.

I'm going to think about this a bit more. Perhaps there are ways Pyright could handle some common cases through the use of code flow analysis. By "common cases" I'm thinking of expressions that involve a simple symbol name (e.g. `x`) with no reassignments of that symbol between the test and the usage. That would cover your example above.

A potential workaround is to use assignment expressions (assuming you're using Python 3.8 or newer).

```
dict_comp = {key: w.strip() if (w := i[key]) is not None else "" for key in i}
```
Currently, Pyright doesn't handle this case as a type guard because of the parentheses around the type assignment. It does handle this case:
```
dict_comp = {key: w.strip() if w := i[key] else "" for key in i}
```
It would be straightforward to add type guard support for parentheses surrounding an assignment expression.

**Comment by falan** [comment]
Thank you very much for your detailed and clear explanation.  I’ve learned a lot from you.

**Comment by erictraut** [comment]
I've added support for the parenthesized assignment expression as mentioned above. Handling the other case is much more involved. I think it's possible, but it adds significant complexity to the type checker that I don't want to add at this time.

**Comment by erictraut** [comment]
This is addressed in Pyright 1.1.41, which I just published.
